### ISSUE TITLE
geom_col switching place unintendedly when position="dodge" and has further aesthetics

### ISSUE THREAD

**Issue by matanhakim** [tidyverse/ggplot2#5217]
Hi, thanks a lot for `{ggplot2}`, it changed my entire workflow for the best!
When using `geom_col` and adding another aesthetic inside it, some columns switch places.
Look at the left pair of columns in each plot: the first plot runs OK, but when the aesthetic is added this pair switches.
notice that `geom_text` doesn't have this problem and the text labels stay where they should be.
This problem first came up using `{ggpattern}` and I filed an [issue there](https://github.com/coolbutuseless/ggpattern/issues/97#issue-1609904165), but as it shows in the reprex here, the bug is in `{ggplot2}` itself.

Here is the code to reproduce the bug:

``` r
library(ggplot2)
#> Warning: package 'ggplot2' was built under R version 4.2.2

# Read the data
df_bug <- structure(list(reason = c("conditions", "conditions", "interest", 
                                      "interest"), year = structure(c(1L, 2L, 1L, 2L), levels = c("2016", 
                                                                                                  "2022"), class = "factor"), pct = c(0.12, 0.33, 0.22, 0.13), 
                         rse = c(0.27, 0.14, 0.19, 0.25), rse_sig = c("medium", "high", 
                                                                      "medium", "medium")), row.names = c(NA, -4L), class = c("tbl_df", 
                                                                                                                              "tbl", "data.frame"))

# This is the tibble
df_bug
#> # A tibble: 4 × 5
#>   reason     year    pct   rse rse_sig
#>   <chr>      <fct> <dbl> <dbl> <chr>  
#> 1 conditions 2016   0.12  0.27 medium 
#> 2 conditions 2022   0.33  0.14 high   
#> 3 interest   2016   0.22  0.19 medium 
#> 4 interest   2022   0.13  0.25 medium

# This works out fine
df_bug |>
  ggplot(aes(reason, pct, fill = year)) +
  geom_col(position = "dodge") +
  geom_text(aes(label = pct), position = position_dodge(width = 1))
```

![](https://i.imgur.com/dlYm3hY.png)

``` r

# This fails when aes(color = rse_sig) is added
df_bug |>
  ggplot(aes(reason, pct, fill = year)) +
  geom_col(aes(color = rse_sig), position = "dodge") +
  geom_text(aes(label = pct), position = position_dodge(width = 1))
```

![](https://i.imgur.com/eRDyh3b.png)

<sup>Created on 2023-03-06 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Thanks a lot!

**Comment by matanhakim** [comment]
I understood that the problem is that the `aes()` is inside `geom_col` and this groups the data by the `rse_sig` variable.
Nevertheless, I didn't see any documentation for this specific kind of behavior.
Is this considered "normal", and if so, where can I read more on this?

**Comment by smouksassi** [comment]
not sure what you are trying to do ?
take ownership of the grouping 

library(ggplot2)
library(patchwork)
df_bug <-
  structure(
    list(
      reason = c("conditions", "conditions", "interest",
                 "interest"),
      year = structure(
        c(1L, 2L, 1L, 2L),
        levels = c("2016",
                   "2022"),
        class = "factor"
      ),
      pct = c(0.12, 0.33, 0.22, 0.13),
      rse = c(0.27, 0.14, 0.19, 0.25),
      rse_sig = c("medium", "high",
                  "medium", "medium")
    ),
    row.names = c(NA,-4L),
    class = c("tbl_df",
              "tbl", "data.frame")
  )

df_bug$rse_sig <- as.factor(df_bug$rse_sig)
df_bug$rse_sig  <- factor(df_bug$rse_sig,levels = c("medium", "high" ))
# This works out fine
p1 <- df_bug |>
  ggplot(aes(reason, pct, fill = year)) +
  geom_col(position = "dodge") +
  geom_text(aes(label = pct), position = position_dodge(width = 1))

p2 <- df_bug |>
  ggplot(aes(reason, pct)) +
  geom_col( aes(fill = year),position = "dodge") +
  geom_text(aes(color = rse_sig,label = pct,group=interaction(rse_sig,reason,year)),
            position = position_dodge(width = 1),vjust=0)
p1 | p2

**Comment by teunbrand** [comment]
Yeah I agree with Samer that this is a grouping issue. If you add `group = interaction(rse_sig, year)` to the main plot's `aes()` call, this should be resolved.

ggplot2 tries to automatically infer grouping based on the mapping of discrete variables (such as fill/colour in your plot). This behaviour is explained in `?aes_group_order`. However, there is only so much inferring to be done when one layer has two grouping variables (`year` and `rse_sig`) and another has not (the text layer).

I agree it is somewhat annoying that you'd have to manually set this, but I don't think there is a reasonable way to infer grouping across different layers with different aesthetics, or which is the 'outer' variable and 'inner' variable (which causes the order of dodging to change). 

I'd like to thank you for you time preparing this report with an excellent reproducible example. Nonetheless, I don't think there is a lot that ggplot2 could do here to improve this situation a lot without breaking a bunch of other assumptions people have been making in their ggplot2 code.
